### PR TITLE
Partial updates

### PR DESCRIPTION
**Pull request by u12206050**

### CODE REVIEW

**Comment by u12206050** [comment]
Here it is. Works even better I believe than the hashCache version.

File: README.md [hunk 1]
```diff
@@ -68,15 +65,39 @@ module.exports = {
...
+
+By default all records will be reindexed on every build. To enable only indexing the new, changed and deleted records include the following in the options of the plugin:
+
+```javascript:title=gatsby-config.js
+  resolve: `gatsby-plugin-algolia`,
+  options: {
+    /* ... */
+    enablePartialUpdates: true,
+    /* (optional) Fields to use for comparing if the index object is different from the new one */
+    /* By default it uses a field called modified which could be a boolean | datatime string */
```
Comment (Haroenv) [inline README.md]: a field called what?

Comment (Haroenv) [reply README.md]: oh I get it now:

```suggestion
    /* By default it uses a field called "modified" which could be a boolean | datetime string */
```

File: README.md [hunk 2]
```diff
@@ -68,15 +65,39 @@ module.exports = {
...
+    matchFields: ['slug', 'modified'] // Array<String> default: ['modified']
+  }
+```
+
+This saves a lot of
```
Comment (Haroenv) [inline README.md]: will need to be finished, of course

File: README.md [hunk 3]
```diff
@@ -55,9 +54,7 @@ const queries = [
     query: myQuery,
     transformer: ({ data }) => data.allSitePage.edges.map(({ node }) => node), // optional
     indexName: 'index name to target', // overrides main index name, optional
-    settings: {
-      // optional, any index settings
-    },
+    matchFields: ['slug', 'modified'], // Array<String> overrides main match fields, optional
```
Comment (Haroenv) [inline README.md]: settings should be put back

File: README.md [hunk 4]
```diff
@@ -17,11 +17,10 @@ First add credentials to a .env file, which you won't commit. If you track this
 ```env
 // .env.production
 ALGOLIA_APP_ID=XXX
-ALGOLIA_API_KEY=XXX
 ALGOLIA_INDEX_NAME=XXX
 ```
 
-```js
+```javascript:title=gatsby-config.js
```
Comment (Haroenv) [inline README.md]: I don't think this syntax works on GitHub

File: README.md [hunk 5]
```diff
@@ -17,11 +17,10 @@ First add credentials to a .env file, which you won't commit. If you track this
 ```env
 // .env.production
 ALGOLIA_APP_ID=XXX
-ALGOLIA_API_KEY=XXX
```
Comment (Haroenv) [inline README.md]: why was this removed?

File: gatsby-node.js [hunk 6]
```diff
@@ -9,59 +9,181 @@ const report = require('gatsby-cli/lib/reporter');
  */
 const identity = obj => obj;
 
+/**
+ * Fetches all records for the current index from Algolia
+ *
+ * @param {AlgoliaIndex} index eg. client.initIndex('your_index_name');
+ * @param {Array<String>} attributesToRetrieve eg. ['modified', 'slug']
+ */
+
```
Comment (Haroenv) [inline gatsby-node.js]:
```suggestion
```

File: package.json [hunk 7]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "gatsby-plugin-algolia",
-  "version": "0.3.0",
+  "version": "0.4.0",
```
Comment (Haroenv) [inline package.json]: this can be done in a separate PR

File: gatsby-node.js [hunk 8]
```diff
@@ -9,59 +9,181 @@ const report = require('gatsby-cli/lib/reporter');
...
+function fetchAlgoliaObjects(index, attributesToRetrieve) {
+  return new Promise((resolve, reject) => {
+    const browser = index.browseAll('', { attributesToRetrieve: ['modified'] });
+    const hits = {};
```
Comment (Haroenv) [inline gatsby-node.js]: a big index can definitely make this go out of memory; maybe a temporary file here that can be written & read as a stream (although that clearly complicates things)

Comment (u12206050) [reply gatsby-node.js]: Since we are only getting 1-3 fields per record it won't be that big of an issue unless it is in the hundreds of thousands of records.

Comment (Haroenv) [reply gatsby-node.js]: let's leave it like this indeed; we can iterate over that later if it's causing issues

File: README.md (hunk 5 above)
Comment (u12206050) [reply README.md]: I don't know, I didn't remove it, I just did a rebase on the master... might be that my master was out of date when I did it.

File: README.md [hunk 9]
```diff
@@ -17,11 +17,11 @@ First add credentials to a .env file, which you won't commit. If you track this
...
+ALGOLIA_APP_KEY=XXX
```
Comment (Haroenv) [inline README.md]:
```suggestion
ALGOLIA_API_KEY=XXX
```

**Comment by teamfphl** [comment]
Looks great. Is there any ETA on when it could be merged? Thanks!

File: README.md (hunk 9 above)
Comment (u12206050) [reply README.md]: done.

File: package.json (hunk 7 above)
Comment (u12206050) [reply package.json]: done.

File: README.md (hunk 5 above)
Comment (u12206050) [reply README.md]: done.

File: README.md (hunk 4 above)
Comment (u12206050) [reply README.md]: done.

**Comment by u12206050** [comment]
Suggestion changes implemented. Ready for merge.

**Comment by Haroenv** [comment]
@teamfphl I'd love for you to try out to verify it works in all kinds of installations, no need to wait until it's released

**Comment by teamfphl** [comment]
Tested and working :)

**Comment by u12206050** [comment]
One month after a successful launch and daily building of the entire site I can confirm this works as it should, both updating only what has changed and removing entries that do not exist anymore.

Would be nice if it is merged into master.

**Comment by Haroenv** [comment]
Perfect, I'm putting it high on my todo list, thanks for your effort @u12206050

File: gatsby-node.js [hunk 10]
```diff
@@ -9,45 +9,126 @@ const report = require('gatsby-cli/lib/reporter');
...
 
+/**
+ * Fetches all records for the current index from Algolia
+ *
+ * @param {AlgoliaIndex} index eg. client.initIndex('your_index_name');
+ * @param {Array<String>} attributesToRetrieve eg. ['modified', 'slug']
+ */
+function fetchAlgoliaObjects(index, attributesToRetrieve) {
+  return new Promise((resolve, reject) => {
+    const browser = index.browseAll('', { attributesToRetrieve: ['modified'] });
```
Comment (MakeNowJust) [inline gatsby-node.js]: Why don't use `attributesToRetrieve`? It seems a bug.

```suggestion
    const browser = index.browseAll('', { attributesToRetrieve });
```

Comment (u12206050) [reply gatsby-node.js]: Thanks, it is actually meant to be the fallback value. Have updated the code.

**Comment by u12206050** [comment]
@Haroenv any reason why this isn't merged yet? I am starting on my third gatsby project with Algolia and I fined I still have to use my own version. Would you merge and publish asap? Otherwise I would have to publish my own version.

**Comment by Haroenv** [comment]
Because I have not had time to use the plugin in the last months, and don't want to merge something I haven't tested

**Comment by fraserisland** [comment]
Can also confirm i'm using this branch in production and have had no issues.

File: gatsby-node.js [hunk 11]
```diff
@@ -9,45 +9,126 @@ const report = require('gatsby-cli/lib/reporter');
...
+      );
+    }
 
-    if (mainIndexExists) {
-      setStatus(activity, `query ${i}: copying existing index`);
-      await scopedCopyIndex(client, index, tmpIndex);
+    const index = client.initIndex(indexName);
+    /* Use temp index if main index already exists */
+    let useTempIndex = false
+    const indexToUse = await (async function(_index) {
```
Comment (bramses) [inline gatsby-node.js]: does this need to be inline? Asking for readability/future updates

File: gatsby-node.js [hunk 12]
```diff
@@ -9,45 +9,126 @@ const report = require('gatsby-cli/lib/reporter');
...
+    setStatus(activity, `query ${i}: graphql resulted in ${Object.keys(objects).length} records`);
+
+    let hasChanged = objects;
+    let algoliaObjects = {}
+    if (enablePartialUpdates) {
+      setStatus(activity, `query ${i}: starting Partial updates`);
+
+      algoliaObjects = await fetchAlgoliaObjects(indexToUse, matchFields);
+
+      const results = Object.keys(algoliaObjects).length
```
Comment (bramses) [inline gatsby-node.js]: suggested rename: `numOfMatchedAlgoliaRecords`

File: gatsby-node.js [hunk 13]
```diff
@@ -59,14 +140,31 @@ exports.onPostBuild = async function(
...
+
+    if (enablePartialUpdates) {
+      /* Execute once per index */
+      /* This allows multiple queries to overlap */
+      const cleanup = Object.keys(indexState).map(async function(indexName) {
+        const state = indexState[indexName];
+        const isRemoved = Object.keys(state.toRemove);
+
+        if (isRemoved.length) {
+          setStatus(activity, `deleting ${isRemoved.length} object from ${indexName} index`);
```
Comment (bramses) [inline gatsby-node.js]: object -> objects

File: gatsby-node.js [hunk 14]
```diff
@@ -59,14 +140,31 @@ exports.onPostBuild = async function(
...
+          const { taskID } = await state.index.deleteObjects(isRemoved);
+          return state.index.waitTask(taskID);
```
Comment (bramses) [inline gatsby-node.js]: can you elaborate on `waitTask()` here? I'm a bit confused :p

File: gatsby-node.js [hunk 15]
```diff
@@ -9,45 +9,126 @@ const report = require('gatsby-cli/lib/reporter');
...
+      toRemove: {}
     }
+    const currentIndexState = indexState[indexName]
 
     setStatus(activity, `query ${i}: executing query`);
     const result = await graphql(query);
     if (result.errors) {
       report.panic(`failed to index to Algolia`, result.errors);
     }
     const objects = transformer(result);
```
Comment (bramses) [inline gatsby-node.js]: Suggested rename: `transformedQueryResults`

File: gatsby-node.js [hunk 16]
```diff
@@ -9,45 +9,126 @@ const report = require('gatsby-cli/lib/reporter');
...
+
+      algoliaObjects = await fetchAlgoliaObjects(indexToUse, matchFields);
+
+      const results = Object.keys(algoliaObjects).length
+      setStatus(activity, `query ${i}: found ${results} existing records`);
+
+      if (results) {
+        hasChanged = objects.filter(curObj => {
+          const ID = curObj.objectID
+          let extObj = algoliaObjects[ID]
```
Comment (bramses) [inline gatsby-node.js]: what if `algoliaObjects[ID]` is null?

File: gatsby-node.js [hunk 17]
```diff
@@ -9,45 +9,126 @@ const report = require('gatsby-cli/lib/reporter');
...
+
+          /* The object exists so we don't need to remove it from Algolia */
+          delete(algoliaObjects[ID]);
+          delete(currentIndexState.toRemove[ID])
+
+          if (!extObj) return true;
+
+          return !!matchFields.find(field => extObj[field] !== curObj[field]);
```
Comment (bramses) [inline gatsby-node.js]: does this do shallow comparison only? IE what happens for nested fields?

File: example/gatsby-config.js [hunk 18]
```diff
@@ -52,6 +53,8 @@ module.exports = {
         indexName: process.env.ALGOLIA_INDEXNAME, // for all queries
         queries,
         chunkSize: 10000, // default: 1000
+        enablePartialUpdates: true // default: true
```
[the remark this answers is not in the document]
Comment (u12206050) [reply example/gatsby-config.js]: Depends on your eslint settings.

File: gatsby-node.js (hunk 16 above)
Comment (u12206050) [reply gatsby-node.js]: it shouldn't be null since it comes from Algolia

File: gatsby-node.js (hunk 17 above)
Comment (u12206050) [reply gatsby-node.js]: Yes shallow comparison only.

File: example/gatsby-config.js (hunk 18 above)
[the remark this answers is not in the document]
Comment (bramses) [reply example/gatsby-config.js]: well its above the `matchFields` key so it's not the last key in the object

File: gatsby-node.js (hunk 17 above)
Comment (bramses) [reply gatsby-node.js]: any potential ideas for nested comparisons? worth mentioning in readme?

**Comment by janosh** [comment]
Very cool! Thanks @u12206050 for this contribution.

@Haroenv Seems like multiple users ([1](https://github.com/algolia/gatsby-plugin-algolia/pull/27#issuecomment-496326582), [2](https://github.com/algolia/gatsby-plugin-algolia/pull/27#issuecomment-495526441), [3](https://github.com/algolia/gatsby-plugin-algolia/pull/27#issuecomment-471756172)) confirmed it's working. Would be great if this was merged!

**Comment by u12206050** [comment]
@janosh  https://www.gatsbyjs.org/packages/gatsby-plugin-algolia-search/

**Comment by ppcano** [comment]
I received this message from algolia.

> You are now incurring over-quota charges according to your plan rates:

> 

> To avoid over-quota charges, we recommend one or more of the following:

> Reduce the number of indexing operations by syncing data less often and/or only re-indexing objects that need to be changed. Learn more about Indexing on Algolia
> Reduce the number of search operations
> Select another plan that better fits your needs

We have a very low number of search operations but we release (re-index) very frequently. 


It looks this PR could help us fixing the over-quota issue by reducing the number of indexing operations as the first option suggests. 

@Haroenv, It's a pity this PR hasn't been merged yet. The feature is really important to avoid `algolia` become expensive very quickly and the PR is relatively small.

**Comment by Haroenv** [comment]
The identified issues haven't been fixed, but I haven't had time myself yet, since I haven't used Gatsby in a while.

@u12206050 has published a fork which uses this behaviour though: https://yarnpkg.com/package/gatsby-plugin-algolia-search

**Comment by ppcano** [comment]
@Haroenv Thank you for the reply. I don't have a problem using that fork, but it gives more confidence to use the package maintained by the team behind the project.

Where do you think it is the best place for requesting the algolia team to work on this issue?

https://discourse.algolia.com/c/feedback

**Comment by u12206050** [comment]
@ppcano We actively maintain the "fork", as we also maintain the same library for Gatsby and Gridsome.
Read more: [Gatsby plugin Algolia with partial updates](https://www.gatsbyjs.org/packages/gatsby-plugin-algolia-search/?=algolia)

**Comment by AnalogMemory** [comment]
@u12206050 

Tried your fork and get this error in console when building. Looks like the `index.js` in the `example/src/layouts` directory is getting picked up by gatsby

```
...
ERROR #85910  GRAPHQL

Multiple "root" queries found: "SiteTitleQuery" and "SiteTitleQuery".
Only the first ("SiteTitleQuery") will be registered.

Instead of:

   1 | query SiteTitleQuery {
   2 |   site {
   3 |     #...
   4 |   }
   5 | }
   6 |
   7 | query SiteTitleQuery {
   8 |   site {
   9 |     #...
  10 |   }
  11 | }

Do:

  1 | query siteTitleQueryAndSiteTitleQuery {
  2 |   site {
  3 |     #...
  4 |   }
  5 |   site {
  6 |     #...
  7 |   }
  8 | }

This can happen when you use two page/static queries in one file. Please combine those into one query.
If you're defining multiple components (each with a static query) in one file, you'll need to move each component to its own file.

File: node_modules/gatsby-plugin-algolia-search/example/src/layouts/index.js
...
```

**Comment by Haroenv** [comment]
I don't think that's related @AnalogMemory, the error message is talking about a query format in layout.js

**Comment by AnalogMemory** [comment]
@Haroenv It's a line in the examples folder, which is getting picked up by Gatsby. 
https://github.com/u12206050/gatsby-plugin-algolia/blob/d76e9c618912cab5164f085c070a905eab10a02f/example/src/layouts/index.js#L38

**Comment by bloudermilk** [comment]
It would be great to get this merged into the official package. Our index only has about ~1k records but we're blowing through our operations quota as our CI pipeline deploys our site many times per day.

It looks like most of the comments in the last PR review were questions or cosmetic suggestions. Maybe one other recent issue that @AnalogMemory reported?

If someone at Algolia can confirm that there is interest in merging I'm happy to pick up the torch.

**Comment by janosh** [comment]
> If someone at Algolia can confirm that there is interest in merging I'm happy to pick up the torch.

That'd be great! There's certainly still interest from my side. 👍

**Comment by ppcano** [comment]
> our CI pipeline deploys our site many times per day.

Same here. 

I open [an issue in discourse.algolia.com](https://discourse.algolia.com/t/gatsby-plugin-overquote/9804)

**Comment by AnalogMemory** [comment]
@bloudermilk I'm using that fork in production and it really saved the day, operations dropped off immediately. The warning doesn't affect the build

Also this is Alex, back in the day when ya'll were at that spot in Venice 👋

**Comment by Haroenv** [comment]
There's interest in getting this merged, but the comments that are left now are to make sure this works in all use cases. I'd love to spend more time on this plugin, but unfortunately Gatsby hasn't been able to be my #1 priority due to some other big projects. I'll definitely review if someone else wants to build on top of this

**Comment by u12206050** [comment]
I have updated my fork and released 0.5.7 without the example folder. If this is at all the case it shouldn't be a problem any more.

Algolia's api does allow for partial updating and that is what my fork makes use of.
**Note** My fork is published on npm as `gatsby-plugin-algolia-search` so make sure you uninstall `gatsby-plugin-algolia` and then `npm install --save gatsby-plugin-algolia-search`

@AnalogMemory, @bloudermilk, @janosh, @ppcano

**Comment by gburgett** [comment]
Thanks @u12206050 !  Saved me a lot of work!

Haroen, if there's anything I can do to help push this along I'm happy to help out.  We've just switched over to using this fork until it gets merged.

File: README.md [hunk 19]
```diff
@@ -68,19 +69,46 @@ module.exports = {
       options: {
         appId: process.env.ALGOLIA_APP_ID,
         apiKey: process.env.ALGOLIA_API_KEY,
-        indexName: process.env.ALGOLIA_INDEX_NAME, // for all queries
+        indexName: "index name to target", // for all queries
```
Comment (Haroenv) [inline README.md:72]:
```suggestion
        indexName: process.env.ALGOLIA_INDEX_NAME, // for all queries
```

File: README.md [hunk 20]
```diff
@@ -68,19 +69,46 @@ module.exports = {
...
+  }
+```
+
+This saves a lot of Algolia operations since you don't reindex everything on everybuild.
+
+### Advanced
+
+You can also specify `matchFields` per query to check for different fields based on the type of objects you are indexing.
+
+The `transformer` field accepts a function and optionally you may provide an `async` function.
```
Comment (Haroenv) [inline README.md:110]:
```suggestion

## Transformer

The `transformer` field accepts a function and optionally you may provide an `async` function. This is useful when you want to change e.g. "edges.node" to simply an array.
```

File: example/gatsby-config.js (hunk 18 above)
[the remark this answers is not in the document]
Comment (Haroenv) [reply example/gatsby-config.js:56]:
```suggestion
        enablePartialUpdates: true, // default: true
```

File: gatsby-node.js [hunk 21]
```diff
@@ -34,26 +61,82 @@ exports.onPostBuild = async function(
...
+    /* Use temp index if main index already exists */
+    let useTempIndex = false
+    const indexToUse = await (async function(_index) {
+      if (!enablePartialUpdates) {
+        if (useTempIndex = await indexExists(_index)) {
+          return client.initIndex(`${indexName}_tmp`);
+        }
+      }
+      return _index
+    })(index)
```
Comment (Haroenv) [inline gatsby-node.js:80]:
```suggestion
    const mainIndexExists = await indexExists(index);
    const indexToUse =
      mainIndexExists && !enablePartialUpdates
        ? client.initIndex(`${indexName}_tmp`)
        : index;
```

File: gatsby-node.js [hunk 22]
```diff
@@ -34,26 +61,82 @@ exports.onPostBuild = async function(
...
 
-    if (mainIndexExists) {
-      setStatus(activity, `query ${i}: copying existing index`);
-      await scopedCopyIndex(client, index, tmpIndex);
+    /* Use to keep track of what to remove afterwards */
+    if (!indexState[indexName]) indexState[indexName] = {
+      index,
+      toRemove: {}
     }
+    const currentIndexState = indexState[indexName]
```
Comment (Haroenv) [inline gatsby-node.js:87]:
```suggestion
    // Use to keep track of what to remove afterwards
    if (!indexState[indexName]) {
      indexState[indexName] = {
        index,
        toRemove: {},
      };
    }

    const currentIndexState = indexState[indexName];
```

File: gatsby-node.js [hunk 23]
```diff
@@ -34,26 +61,82 @@ exports.onPostBuild = async function(
...
+        if (useTempIndex = await indexExists(_index)) {
+          return client.initIndex(`${indexName}_tmp`);
+        }
+      }
+      return _index
+    })(index)
 
-    if (mainIndexExists) {
-      setStatus(activity, `query ${i}: copying existing index`);
-      await scopedCopyIndex(client, index, tmpIndex);
```
Comment (Haroenv) [inline gatsby-node.js:44]: this code is gone now, I think that means that the enablePartialUpdates: false mode no longer works. This should be tested :) when the main index already exists

File: gatsby-node.js (hunk 13 above)
Comment (Haroenv) [reply gatsby-node.js:179]:
```suggestion
          setStatus(activity, `deleting ${isRemoved.length} objects from ${indexName} index`);
```

**Comment by Haroenv** [comment]
Thanks for suggesting @gburgett, The questions that still need to be answered are:

- does this still work with the partial updates disabled (especially if the main index already exists)
- suggestions which are left inline, which are mostly superficial

**Comment by gburgett** [comment]
Sure, happy to test locally with our workflow!  We use a "digest" field on our algolia documents to detect if anything has changed.  Our configuration looks like this:

```
gatsbyConfig.plugins.push({
    resolve: `gatsby-plugin-algolia-search`,
    options: {
      appId: process.env.GATSBY_ALGOLIA_APP_ID,
      apiKey: process.env.ALGOLIA_API_KEY,
      indexName: process.env.GATSBY_ALGOLIA_INDEX_NAME, // for all queries
      queries: algolia.queries,
      chunkSize: 10000, // default: 1000,
      enablePartialUpdates: true,
      matchFields: ['digest'],
    },
  })
```

Here's what happens when I run with partial updates on my ~1500 blog posts:

```
[Algolia Plugin] 1 queries to index - 0.013 ms
[Algolia Plugin] query 0: executing query - 0 ms
[Algolia Plugin] query 0: graphql resulted in 1468 records - 1.883 ms
[Algolia Plugin] query 0: starting Partial updates - 0 ms
[Algolia Plugin] query 0: found 1476 existing records - 0.502 ms
[Algolia Plugin] query 0: Partial updates – [insert/update: 70, total: 1468] - 0.001 ms
[Algolia Plugin] query 0: splitting in 1 jobs - 0.001 ms
[Algolia Plugin] deleting 8 object from papyrus_development_blogs index - 6.457 ms
[Algolia Plugin] Done in 14.739 s
```

This generates two batch updates
image
image

If I set `enablePartialUpdates: false` and rebuild:

```
[Algolia Plugin] 1 queries to index - 0.011 ms
[Algolia Plugin] query 0: executing query - 0.384 ms
[Algolia Plugin] query 0: graphql resulted in 1468 records - 1.975 ms
[Algolia Plugin] query 0: splitting in 1 jobs - 0.001 ms
[Algolia Plugin] query 0: moving copied index to main index - 6.084 ms
[Algolia Plugin] Done in 9.159 s
```
And I see these lines in the Algolia logs:
image
image

My index still has the correct  1,468 records

**Comment by Haroenv** [comment]
I'm sorry for being so slow on this pr, but this is now released on 0.8.0

**Comment by janosh** [comment]
@Haroenv No problem, just happy it got merged. 😄

**Comment by bloudermilk** [comment]
Huge thanks to everyone who stepped up to get this shipped.
